src/eddymotion/model/base.py

```diff
@@ -409,6 +412,119 @@ class DKIModel(BaseModel):
+    3. Compute for M iterations
+     - Load GP prediction maker
+     - For all i in N (DWIs)
+       - Compute \\hat{s}_{i} (f_{i}, h, \beta_{i}, r_{i}, a_{i}) eqs 2 and 4
+       - Load \\hat{s}_{i} (f_{i}, h, \beta_{i}, r_{i}, a_{i}) as training
+         data for GP
+     - Estimate hyperparameters for the GP used to predict the signal shape
+       for every voxel
+     - Update EC and movement parameters
+     - For all i in N (DWIs)
+       - Draw a prediction s_{i} from the GP
+       - Compute \\hat{f}_{i} (s_{i}, h, \beta_{i}, r_{i}, a_{i})
+       - Use \\hat{f}_{i} - f_{i} to update \beta_{i} and r_{i} (eq 6)
+
+    a: direction of the PE and the total readout time (here defined as the time
+    between the acquisition of the center of the first and last echoes).
+    Internally, a is divided into a = [p t] where p is a unity length 1 x 3
+    vector defining the PE direction (such that for example [1 0 0], [−1 0 0],
+    [0 1 0] and [0 −1 0] denote R → L, L → R, P → A and A  P PE direction
+    respectively) and where t denotes the readout time (in seconds).
+    r: 1x6 vector: 3 translations, 3 rotations
+    \beta: four for linear; ten for quadratic, twenty for cubic.
+    h: assumed to be in the same space as the first b = 0 image supplied to
+    eddy, which will be automatically fulfilled if it was estimated by topup
+    and that same b = 0 image was the first of those supplied to topup. Hence,
+    it can be said to help define the reference/undistorted space as the first
+    b = 0 image after distortion correction by h.
+
+    See Appendix A for further details.
+
+    Add the outlier detection part in [Andersson16b]?
+
+    References
+    ----------
+    .. [Andersson16a] J. L. R. Andersson. et al., An integrated approach to
+    correction for off-resonance effects and subject movement in diffusion MR
+    imaging, NeuroImage 125 (2016) 1063–1078
+    .. [Andersson16b] J. L. R. Andersson. et al., Incorporating outlier
+    detection and replacement into a non-parametric framework for movement and
+    distortion correction of diffusion MR images, NeuroImage 141 (2016) 556–572
+    """
+
+    __slots__ = (
+        "_dwi",
+        "_a",
+        "_h",
+        "_kernel",
+        "_num_iterations",
+        "_betas",
+        "_r",
+        "_gpr",
+        "_model",
+    )
+
+    def __init__(self, dwi, a, h, kernel, num_iterations=5, **kwargs):
+        """Implement object initialization."""
+
+        self._dwi = dwi  # The HDF5 file object: avoid having the entire 4D volume in memory
+        self._a = a
+        self._h = h
+        self._num_iterations = num_iterations
+
+        # Initialize
+        self._betas = 0
+        self._r = 0
+
+        # ToDo
+        # Build the GP kernel here or in fit ?
+        self._gpr = None
+        # Does the kernel depend on which data we use as the training data (i.e.
+        # varies with the index we choose to predict)?
+        self._kernel = kernel
```

Comment:
I don't think it does. I believe we just need to keep the kernel between iterations.
```suggestion
        self._kernel = kernel
```